### PR TITLE
Fix for before/after test listeners not failing tests

### LINKED ISSUE #771: Error handling in tests
Hello!

I found a really serious bug on the test runner, when there is an exception in some test.

The effect is really hard to detect.

I have this stacktrace:

```
17:18:32.856 [main] ERROR io.kotlintest.runner.jvm.TestEngine - Error during test engine run
java.lang.reflect.InvocationTargetException: null
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_144]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_144]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_144]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_144]
	at kotlin.reflect.jvm.internal.FunctionCaller$Constructor.call(FunctionCaller.kt:66) ~[kotlin-reflect-1.2.51.jar:1.2.51-release-125 (1.2.51)]
	at kotlin.reflect.jvm.internal.KCallableImpl.call(KCallableImpl.kt:107) ~[kotlin-reflect-1.2.51.jar:1.2.51-release-125 (1.2.51)]
	at kotlin.reflect.jvm.internal.KCallableImpl.callDefaultMethod(KCallableImpl.kt:149) ~[kotlin-reflect-1.2.51.jar:1.2.51-release-125 (1.2.51)]
	at kotlin.reflect.jvm.internal.KCallableImpl.callBy(KCallableImpl.kt:111) ~[kotlin-reflect-1.2.51.jar:1.2.51-release-125 (1.2.51)]
	at kotlin.reflect.full.KClasses.createInstance(KClasses.kt:283) ~[kotlin-reflect-1.2.51.jar:1.2.51-release-125 (1.2.51)]
	at io.kotlintest.runner.jvm.JvmKt.instantiateSpec(jvm.kt:15) ~[kotlintest-runner-jvm-3.3.2.jar:na]
	at io.kotlintest.runner.jvm.TestEngine.createSpec(TestEngine.kt:112) ~[kotlintest-runner-jvm-3.3.2.jar:na]
	at io.kotlintest.runner.jvm.TestEngine.access$createSpec(TestEngine.kt:15) ~[kotlintest-runner-jvm-3.3.2.jar:na]
	at io.kotlintest.runner.jvm.TestEngine$submitSpec$1.run(TestEngine.kt:91) ~[kotlintest-runner-jvm-3.3.2.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_144]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_144]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_144]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_144]
Caused by: java.lang.NullPointerException: null
```

The problem is that this error is affecting other tests.

I think that kotlintest is ordering the tests to run alphabetically by name, so, when the engine throws this exception, it doesn't continue with any other test, and it doesn't inform that it failed in any way.

image

It just says that all the test runned fine, and the only way to notice the error, is by noticing that the number of total tests is not what it should be.

So, I think that there are two problems here:

1- If there is an exception in a test, it shouldn't break everything
2- If for any reason the engine fails with an exception, it should be informed in the interface, and it should mark the tests as failed, showing it in maven or gradle.

The bug it's really easy to reproduce, just find the first test that should run alphabetically (or add an AAA as prefix in any test) and put a throw NullPointerException()

In that case, it won't run any test, and it will just say that no test were found.
If you do that for the second test, it will run the first one, and don't say anything about the rest.

The worst about this problem, is that it's really hard to notice that something is failing, so you could have a lot of errors in production believing that all your tests are ok.

### PR DESCRIPTION
**Pull request by sksamuel**
This is the second part of the fix for #771.
My previous PR fixed before/after spec.
This fixes before/after test.

closes #771

### CODE REVIEW

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt [hunk 1]
```diff
@@ -48,7 +48,7 @@ class BigIntegerGen(maxNumBits: Int) : Gen<BigInteger> {
  * from across the entire integer range.
  */
 @higherkind
-interface Gen<T> : GenOf<T> {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/properties/Gen.kt:51]: Why did we remove this integration?

Was this an accident?

File: kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/ConfigTest.kt [hunk 2]
```diff
@@ -33,7 +33,7 @@ class ConfigTest : WordSpec() {
         fail("shouldn't run")
       }
 
-      // If we have 100 threads, and each one sleeps for 100 milliseconds, then the total time
+      // If we have 100 threads, and each one sle`eps for 100 milliseconds, then the total time
```
Comment (LeoColman) [inline kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/ConfigTest.kt:36]: I'm pretty sure this isn't a fix xD

File: kotlintest-tests/kotlintest-tests-junit5/src/test/kotlin/com/sksamuel/kotlintest/junit5/StringSpecExceptionInBeforeTest.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.sksamuel.kotlintest.junit5
+
+import io.kotlintest.TestCase
+import io.kotlintest.shouldBe
+import io.kotlintest.specs.StringSpec
+
+class StringSpecExceptionInBeforeTest : StringSpec() {
```
Comment (LeoColman) [inline kotlintest-tests/kotlintest-tests-junit5/src/test/kotlin/com/sksamuel/kotlintest/junit5/StringSpecExceptionInBeforeTest.kt:7]: How does this test not fail?
Is it not executed in CI?

**Review by LeoColman — APPROVED** [review]

**Comment by LeoColman** [comment]
Code is not compiling due to arrow `GenOf` removal. Idk if it was intentional for this commit.

**Comment by sksamuel** [comment]
I've checked in a bunch of stuff I shouldn't have

**Comment by sksamuel** [comment]
Done this without the extra fluff in #865

**Pull request closed without merging**